### ISSUE TITLE
Activation of the Python extension

### ISSUE THREAD

**Issue by kieferrm** [microsoft/vscode-jupyter#5193]
Issue Type: <b>Bug</b>

When I startup VS Code I get this in a workspace that exclusively contains markdown files. This is the only window I have open. Why?

![image](https://user-images.githubusercontent.com/4674940/110877623-84ce1f80-828e-11eb-896b-c9d1db62f540.png)

Extension version: 2021.2.636928669
VS Code version: Code - Insiders 1.55.0-insider (Universal) (9b2ee7fb7d970b9a628fdb3545bdd01d67078d16, 2021-03-11T05:15:23.749Z)
OS version: Darwin arm64 20.3.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|undefined|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 3, 2|
|Memory (System)|16.00GB (0.28GB free)|
|Process Argv|--enable-proposed-api lszomoru.lszomoru-proposed-api-sample --crash-reporter-id ef10dfea-3602-4c74-ba1a-109ce7da7825|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368:30146709
vsreu685:30147344
python383:30185418
pythonvspyt602:30263608
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstry244:30244315
pythonvsdeb440:30224570
pythonvsded773:30223139
pythonvspyt875:30259475
pythonvsnew554:30265444
pythontb:30258533
vscoresta800:30265947
vspre833:30267464

```

</details>

**Comment by greazer** [comment]
Looks like a dupe of #4965

**Comment by rchiodo** [comment]
This is a dependency issue. Because the typescript notebook extension loads on startup, it also loads the jupyter extension.

**Comment by greazer** [comment]
Ah yes, the other bug was about installing *Python*, sorry. However, the jupyter extension should not need to load the Python extension. The dependency isn't supposed to work that way.

**Comment by greazer** [comment]
Dupe of #4526

**Comment by rchiodo** [comment]
Not a dupe of #4526.

**Comment by rchiodo** [comment]
The root cause of this is here:

https://github.com/microsoft/vscode-jupyter/blob/d99989e17bc4174dba3e865954472d63df01ff1d/src/client/datascience/serviceRegistry.ts#L220

Maybe there's another way we can determine if the python extension is installed other than calling the getExtension API.

**Comment by DonJayamanne** [comment]
I think this is how it has always worked, looks like just because the extension is installed we assume it has been actiavted and try to use it (when we try to use this, then we actiavte it if hasn't been activated).

**Comment by DonJayamanne** [comment]
When i open a typescript notebook, we still activate Python extension (if it is installed) because we're searching for kernels in interpreters.
Not sure i can get around this.
Basically we don't know what kind of notebook the user is dealing with. 

**E.g. assume the following**
* User has python extension installed
* User creates a blank notebook
* Now user goes into kernel picker:
	* How do we know whther user is creating python notebook or not.
	* We need to display kernels in Python interpreters.
	

I.e. i don't see any way around this.
Thoughts?

**Comment by rchiodo** [comment]
Notebook seems okay to load the python extension. It just shouldn't load in a directory with markdown files.

**Comment by greazer** [comment]
One way we could potentially work around this problem is to prompt the user to set a language for the first cell of their notebook before the kernel picker is populated with anything. Right? 

Example: Assuming your scenario, the user has a blank notebook with no language set in the first cell. 
- User click on the kernel picker. They just see a prompt describing the situation. Could be modal or not modal. Or perhaps the popup for the kernel picker just says "Choose a language for the first cell to choose a kernel"

**Comment by IanMatthewHuff** [comment]
I dunno, still seems like if I was a Julia or R user that I wouldn't want the python extension activated just for opening a notebook.  Creating a blank notebook could save a default language to use for new notebooks. Then only activate and search for python language notebooks?

**Comment by DonJayamanne** [comment]
> User click on the kernel picker. They just see a prompt describing the situation. Could be modal or not modal. Or perhaps the popup for the kernel picker just says "Choose a language for the first cell to choose a kernel"

Not sure i like this.
I'm with @rchiodo, its ok for notebooks.
I.e. my argument is, if the user opens a notebook and they have python extension installed, then I its ok to activate Python extension.


Having to get the user to first select a language then the kernel feels like too many steps.

**Comment by DonJayamanne** [comment]
> Creating a blank notebook could save a default language to use for new notebooks. Then only activate and search for python language notebooks?

Sure, but what if i now want to change the kernel to a Python kernel? At this point we won't be displaying all the python interpreters as the Python extension hasn't been actiavted.

I think its ok to activate the python extension.
The crux of the problem is Python extnesion is displaying prompts upon activation and fixing that would not cause issues. I.e. if we didn't have popups would this be an issue, i don't think so.

I'd say, if user installed Python extnesion, then its ok to activate.

**Comment by DonJayamanne** [comment]
Python extenion should display prompts like `select a pytohn interpreetr` only when user attempts to do somethign with python, editing a file, or running some python command that requires python.

Else I don't see the need for the prompt just because the extension actiavted. But that's an issue at their end.

**Comment by greazer** [comment]
Ok. I think I see what you're saying Don. Here's a summary...
- If the user does not have the Python extension installed and a user opens a blank or non-Python notebook, then they will not be prompted to install the Python extension (let alone have it activate) Question: what will the user see if they click on the kernel picker in this state?
- If the user does not have the Python extension installed and a user opens a Python notebook OR the user chooses Python as the language for the first cell, they WILL be prompted to install the Python extension.
- If the user has the Python extension installed and a user opens a non-Python notebook, then the Python extension will be activated (as silently and quickly as possible) such that we can show a list of Python interpreters/kernels in addition to other jupyter kernels we find when the user clicks on the kernel picker.
- If the user has the Python extension installed and a user opens a Python notebook, then the Python extension is activated and kernel picker displays everything as we normally do today.

Is this complete? Do I have it right? Not sure I think it's the right answer yet, though.

**Comment by DonJayamanne** [comment]
> If the user does not have the Python extension installed and a user opens a blank or non-Python notebook, then they will not be prompted to install the Python extension (let alone have it activate) Question: what will the user see if they click on the kernel picker in this state?

They will see the list of the kernels 

> If the user does not have the Python extension installed and a user opens a Python notebook OR the user chooses Python as the language for the first cell, they WILL be prompted to install the Python extension.

Yes (but not beacuse the langauge of the first cell is python, that has no bearing on this).

> If the user has the Python extension installed and a user opens a non-Python notebook, then the Python extension will be activated (as silently and quickly as possible) such that we can show a list of Python interpreters/kernels in addition to other jupyter kernels we find when the user clicks on the kernel picker.

Yes,
But ther'es no such thing as `(as silently and quickly as possible)`.
**This is the problem**

> If the user has the Python extension installed and a user opens a Python notebook, then the Python extension is activated and kernel picker displays everything as we normally do today.


Yes

**Comment by greazer** [comment]
Yes, I agree that the silently and quickly as possible is the main problem here (which is why mentioned it). :D 

I don't think it's a good user experience to open a Julia notebook and even see the Python extension loading. Especially if they have to wait for it to finish to do whatever they want to do next. Thinking outside the box, what options do we have to aleviate this? I'm sure there's SOMETHING that can be done, whether we do it or not we can answer later.

**Comment by DonJayamanne** [comment]
> Thinking outside the box, what options do we have to aleviate this? I'm sure there's SOMETHING that can be done, whether we do it or not we can answer later.

2 suggestions:
* Python extension split environment related functionality into a seprate extension & we can depend on that & use that (that extension doesn't have any UI, only provides API).
* Python extension removes the prompt that gets displayed (this doesn't prevent the extension from loading, but at least users won't get popups and won't notice this as much)


@brettcannon @luabud /cc

**Comment by greazer** [comment]
Thanks Don, my understanding is that option #1 is something that the Python team needed a good excuse to do. They have one now. Note that I don't think it's absolutely required that this API be provided in the form of an extension. But that would be preferable. 

Users will still see an extension being activated, but it could be named appropriately to reflect what it's doing (and thus better imply why it's needed). 

@karthiknadig /cc

**Comment by luabud** [comment]
I'd strongly prefer for it not to be a separate extension. I believe there are other ways we can solve that (e.g. through an npm package?).

My main concern regarding the separate extension is that it IMO adds unnecessary complexity to the UX of installation of the Python extension.

Suppose we did have a "Python Environments" extension. We'd need other functionality like the start page,  start REPL and things like that be somewhere else. What would we call that extension, and how would users understand how that is different from the Python environments one?
We have some users complaining about the extra entry for the Jupyter extension in their installed extensions view, I imagine a lot of people would be annoyed and confused about this extra one too.

I can see us breaking the extension down into formatters and linters and even having a pytest extension instead of a built-in solution.  But I'm struggling to see the user value in having a "Python Environments" and a "Python core" extension.

**Comment by greazer** [comment]
I don't think this needed to be exposed as an extension. But an npm package is still something the Python core team needs to provide right? Assuming that I'm copying this to your repo.

**Comment by brettcannon** [comment]
I wouldn't say "need to" if you're asking if _we_ need it. We _can_ and we originally _wanted_ to, but "need" is too strong of a statement (at least if it was coming out of my mouth). Now if "need" is from the Jupyter extension, then yes, the npm package would be something we would provide to you.

**Comment by greazer** [comment]
:) By "needs" I meant the last thing you said, that's all.

**Comment by DonJayamanne** [comment]
This has been done, but we need tests for this.

**Comment by DonJayamanne** [comment]
> his has been done, but we need tests for this.

We've regressed as a result of the changes for Kernel Push (tests really need to be added for this so we don't regress).

**Comment by greazer** [comment]
It seems ok to activate the python extension if there are python cells in it. This just needs to e thought through with our getting started investigation.

**Comment by rchiodo** [comment]
I believe the test would be:

- create untitled notebook
- verify python extension does not load

But how does it create the notebook controllers then?

**Comment by rchiodo** [comment]
Actually that seems like it should load the python extension. As long as it doesn't prevent opening (which it currently doesn't).

I don't think there's a scenario to test for here then.
